flyteplugins/go/tasks/pluginmachinery/ioutils/remote_file_output_reader.go

```diff
@@ -3,75 +3,243 @@ package ioutils
+		},
+		errorFilePath: errorFilePath,
 	}
+}
+
+func (b *baseErrorRetriever) validatePayloadSize(metadata storage.Metadata) error {
 	if metadata.Exists() {
-		if metadata.Size() > r.maxPayloadSize {
-			return false, errors.Wrapf(err, "error file @[%s] is too large [%d] bytes, max allowed [%d] bytes", r.outPath.GetErrorPath(), metadata.Size(), r.maxPayloadSize)
+		if metadata.Size() > b.maxPayloadSize {
+			return errors.Errorf("file is too large [%d] bytes, max allowed [%d] bytes", metadata.Size(), b.maxPayloadSize)
 		}
-		return true, nil
 	}
-	return false, nil
+	return nil
 }
 
-func (r RemoteFileOutputReader) ReadError(ctx context.Context) (io.ExecutionError, error) {
-	errorDoc := &core.ErrorDocument{}
-	err := r.store.ReadProtobuf(ctx, r.outPath.GetErrorPath(), errorDoc)
+func (s *SingleFileErrorRetriever) HasError(ctx context.Context) (bool, error) {
+	metadata, err := s.store.Head(ctx, s.errorFilePath)
 	if err != nil {
-		if storage.IsNotFound(err) {
-			return io.ExecutionError{
-				IsRecoverable: true,
-				ExecutionError: &core.ExecutionError{
-					Code:    "ErrorFileNotFound",
-					Message: err.Error(),
-					Kind:    core.ExecutionError_SYSTEM,
-				},
-			}, nil
-		}
-		return io.ExecutionError{}, errors.Wrapf(err, "failed to read error data from task @[%s]", r.outPath.GetErrorPath())
+		return false, errors.Wrapf(err, "failed to read error file @[%s]", s.errorFilePath)
+	}
+	err = s.validatePayloadSize(metadata)
+	if err != nil {
+		return false, errors.Wrapf(err, "error file @[%s] failed payload size validation", s.errorFilePath)
 	}
+	return metadata.Exists(), nil
+}
 
+func errorDoc2ExecutionError(errorDoc *core.ErrorDocument, errorFilePath storage.DataReference) io.ExecutionError {
 	if errorDoc.Error == nil {
 		return io.ExecutionError{
 			IsRecoverable: true,
 			ExecutionError: &core.ExecutionError{
 				Code:    "ErrorFileBadFormat",
-				Message: fmt.Sprintf("error not formatted correctly, nil error @path [%s]", r.outPath.GetErrorPath()),
+				Message: fmt.Sprintf("error not formatted correctly, nil error @path [%s]", errorFilePath),
 				Kind:    core.ExecutionError_SYSTEM,
 			},
-		}, nil
+		}
 	}
-
-	ee := io.ExecutionError{
+	executionError := io.ExecutionError{
 		ExecutionError: &core.ExecutionError{
 			Code:    errorDoc.Error.Code,
 			Message: errorDoc.Error.Message,
 			Kind:    errorDoc.Error.Origin,
+			Worker:  errorDoc.Error.Worker,
 		},
 	}
 
 	if errorDoc.Error.Kind == core.ContainerError_RECOVERABLE {
-		ee.IsRecoverable = true
+		executionError.IsRecoverable = true
+	}
+
+	return executionError
+}
+
+func (s *SingleFileErrorRetriever) GetError(ctx context.Context) (io.ExecutionError, error) {
+	errorDoc := &core.ErrorDocument{}
+	err := s.store.ReadProtobuf(ctx, s.errorFilePath, errorDoc)
+	if err != nil {
+		if storage.IsNotFound(err) {
+			return io.ExecutionError{
+				IsRecoverable: true,
+				ExecutionError: &core.ExecutionError{
+					Code:    "ErrorFileNotFound",
```

Comment:
Move to a const?

Reply:
Just moved this from existing code, but makes sense to move into a constant. Will do.

Reply:
Addressed.